### PR TITLE
UCT/DC: Create DCI via DEVX

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5.c [hunk 1]
```diff
@@ -285,6 +286,19 @@ static ucs_status_t uct_dc_mlx5_iface_create_qp(uct_dc_mlx5_iface_t *iface,
     uct_rc_mlx5_iface_fill_attr(&iface->super, &attr,
                                 iface->super.super.config.tx_qp_len,
                                 &iface->super.rx.srq);
+
+    if (md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_DCI) {
+        attr.mmio_mode                        = iface->super.tx.mmio_mode;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:291]: can uct_rc_mlx5_iface_fill_attr set mmio_mode ?

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -522,9 +538,10 @@ void uct_dc_mlx5_destroy_dct(uct_dc_mlx5_iface_t *iface)
 
 static void uct_dc_mlx5_iface_cleanup_dcis(uct_dc_mlx5_iface_t *iface)
 {
+    int max = uct_dc_mlx5_iface_total_ndci(iface);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:541]: max -> num_dcis

File: src/uct/ib/mlx5/dv/ib_mlx5_dv.c [hunk 3]
```diff
@@ -155,20 +163,22 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
 
     qp->qp_num = UCT_IB_MLX5DV_GET(create_qp_out, out, qpn);
 
-    qpc = UCT_IB_MLX5DV_ADDR_OF(rst2init_qp_in, in_2init, qpc);
-    UCT_IB_MLX5DV_SET(rst2init_qp_in, in_2init, opcode, UCT_IB_MLX5_CMD_OP_RST2INIT_QP);
-    UCT_IB_MLX5DV_SET(rst2init_qp_in, in_2init, qpn, qp->qp_num);
-    UCT_IB_MLX5DV_SET(qpc, qpc, pm_state, UCT_IB_MLX5_QPC_PM_STATE_MIGRATED);
-    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.vhca_port_num, attr->super.port);
-    UCT_IB_MLX5DV_SET(qpc, qpc, rwe, true);
+    if (attr->super.qp_type != UCT_IB_QPT_DCI) {
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5_dv.c:166]: == IBV_QPT_RC

File: src/uct/ib/mlx5/dv/ib_mlx5_dv.c [hunk 4]
```diff
@@ -117,7 +117,15 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
 
     UCT_IB_MLX5DV_SET(create_qp_in, in, opcode, UCT_IB_MLX5_CMD_OP_CREATE_QP);
     qpc = UCT_IB_MLX5DV_ADDR_OF(create_qp_in, in, qpc);
-    UCT_IB_MLX5DV_SET(qpc, qpc, st, UCT_IB_MLX5_QPC_ST_RC);
+    if (attr->super.qp_type == UCT_IB_QPT_DCI) {
```
Comment (brminich) [inline src/uct/ib/mlx5/dv/ib_mlx5_dv.c:120]: Seems you also have to pass `dc==1` to ` uct_ib_mlx5_qpc_cs_res(attr->super.max_inl_cqe[UCT_IB_DIR_RX], 0)` on line 143 below. Maybe in this case you will not need set `attr.super.max_inl_cqe[UCT_IB_DIR_RX] = 0;` in `uct_dc_mlx5_iface_create_dci`?

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 5]
```diff
@@ -179,6 +179,7 @@ enum {
     UCT_IB_MLX5_MD_FLAG_DEVX_RC_SRQ      = UCT_IB_MLX5_MD_FLAG_DEVX_OBJS(RCSRQ),
     UCT_IB_MLX5_MD_FLAG_DEVX_DCT         = UCT_IB_MLX5_MD_FLAG_DEVX_OBJS(DCT),
     UCT_IB_MLX5_MD_FLAG_DEVX_DC_SRQ      = UCT_IB_MLX5_MD_FLAG_DEVX_OBJS(DCSRQ),
+    UCT_IB_MLX5_MD_FLAG_DEVX_DCI         = UCT_IB_MLX5_MD_FLAG_DEVX_OBJS(DCI),
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5.h:182]: no need for trailing comma

File: src/uct/ib/mlx5/dv/ib_mlx5_dv.c (hunk 4 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/dv/ib_mlx5_dv.c:120]: nope, `dc=1` actually means DCT, i.e. either 64 or 0. In DCI it should be 0.

Comment (brminich) [reply src/uct/ib/mlx5/dv/ib_mlx5_dv.c:120]: ok

**Review by yosefe — APPROVED** [review]
pls squash

**Review by brminich — APPROVED** [review]

**Pull request merged**